cupy/random/cupy_distributions.cu

```diff
@@ -0,0 +1,211 @@
+    }
+}
+
+__device__ double rk_beta(rk_state* state, double a, double b) {
+    double Ga, Gb;
+    if ((a <= 1.0) && (b <= 1.0)) {
+        double U, V, X, Y;
+        /* Use Johnk's algorithm */
+        while (1) {
+            U = state->rk_double();
+            V = state->rk_double();
+            X = pow(U, 1.0/a);
+            Y = pow(V, 1.0/b);
+            if ((X + Y) <= 1.0) {
+                if (X +Y > 0) {
+                    return X / (X + Y);
+                } else {
+                    double logX = log(U) / a;
+                    double logY = log(V) / b;
+                    double logM = logX > logY ? logX : logY;
+                    logX -= logM;
+                    logY -= logM;
+                    return exp(logX - log(exp(logX) + exp(logY)));
+                }
+            }
+        }
+    } else {
+        Ga = rk_standard_gamma(state, a);
+        Gb = rk_standard_gamma(state, b);
+        return Ga/(Ga + Gb);
+    }
+}
+
+
+__device__ uint32_t rk_interval_32(rk_state* state, int mx, int mask) {
+    uint32_t sampled = state->rk_int() & mask;
+    while(sampled > mx)  {
+        sampled = state->rk_int() & mask;
+    }
+    return sampled;
+}
+
+__device__ uint64_t rk_interval_64(rk_state* state, uint64_t  mx, uint64_t mask) {
+    uint32_t hi= state->rk_int();
+    uint32_t lo= state->rk_int();
+    uint64_t sampled = (static_cast<uint64_t>(hi) << 32 | lo)  & mask;
+    while(sampled > mx)  {
+        hi= state->rk_int();
+        lo= state->rk_int();
+        sampled = (static_cast<uint64_t>(hi) << 32 | lo) & mask;
+    }
+    return sampled;
+}
+
+
+struct interval_32_functor {
+    template<typename... Args>
+    __device__ uint32_t operator () (Args&&... args) {
+        return rk_interval_32(args...);
+    }
+};
+
+struct interval_64_functor {
+    template<typename... Args>
+    __device__ uint64_t operator () (Args&&... args) {
+        return rk_interval_64(args...);
+    }
+};
+
+struct beta_functor {
+    template<typename... Args>
+    __device__ double operator () (Args&&... args) {
+        return rk_beta(args...);
+    }
+};
+
+// There are several errors when trying to do this a full template
+struct exponential_functor {
+    template<typename... Args>
+    __device__ double operator () (Args&&... args) {
+        return rk_standard_exponential(args...);
+    }
+};
+
+template<typename F, typename T, typename R, typename... Args>
+__global__ void execute_dist(intptr_t param, void* out, ssize_t size, Args... args) {
+    int id = threadIdx.x + blockIdx.x * blockDim.x;
+    if (id < size) {
+        T random;
+        F func;
+        random.init_state(id, param);
+        ((R*) out)[id] = func(&random, args...);
+    }
+    return;
+}
+
+void interval_32(intptr_t param, int mx, int mask, void* out, ssize_t size) {
+    int tpb = 256;
+    int bpg =  (size + tpb - 1) / tpb;
+    execute_dist<interval_32_functor, curand_pseudorand_state, int32_t><<<bpg, tpb>>>(param, out, size, mx, mask);
```

Comment:
These are entry points from cython, I am still wondering how to generalize them for different random generators without using an enum and tons of ifs. But with templates, I couldn't get Cython to get the right C object to call the correct template. Any clue about this @asi1024 @leofang ?

Reply:
Is this issue already resolved?

Reply:
yes, by using jitify we could solve this :)